File: tourism-be/src/Modules/Encounters/Explorer.Encounters.Core/Domain/Challenge.cs
using Explorer.BuildingBlocks.Core.Domain;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Explorer.Encounters.Core.Domain
{
    public class Challenge : Entity
    {
        public long CreatorId { get; init; }
        public string Description { get; init; }
        public string Name { get; init; }
        public ChallengeStatus Status { get; init; }
        public ChallengeType Type { get; init; }
        public double Latitude { get; init; }
        public double Longitude { get; init; }
        public int ExperiencePoints { get; init; }
        public long? KeyPointId { get; init; }
        public Uri? Image { get; init; }
        public double? LatitudeImage { get; init; }
        public double? LongitudeImage { get; init; }
        public double Range { get; init; }
        public int? RequiredAttendance { get; set; }

        public Challenge(long creatorId, string description, string name, ChallengeStatus status, ChallengeType type, double latitude, double longitude, 
            Uri? image, double? latitudeImage, double? longitudeImage, double range, int experiencePoints, long? keyPointId, int? requiredAttendance)
        {
            CreatorId = creatorId;
            Description = description;
            Name = name;
            Status = status;
            Type = type;
            Latitude = latitude;
            Longitude = longitude;
            Image = image;
            LatitudeImage = latitudeImage;
            LongitudeImage = longitudeImage;
            Range = range;
            ExperiencePoints = experiencePoints;
            KeyPointId = keyPointId;
            RequiredAttendance = requiredAttendance;

            Validate();
        }
        private void Validate()
        {
            if (string.IsNullOrWhiteSpace(Description)) throw new ArgumentException("Invalid Description");
            if (string.IsNullOrWhiteSpace(Name)) throw new ArgumentException("Invalid Name");
            if (Latitude is > 90 or < -90) throw new ArgumentException("Invalid latitude");
            if (Longitude is > 180 or < -180) throw new ArgumentException("Invalid longitude");
        }
    }
    public enum ChallengeStatus
    {
        Draft,
        Active,
        Archived
    }

    public enum ChallengeType
    {
        Social,
        Location,
        Misc
    }
}

File: golang-be/Encounters/model/Challenge.go
package model

import (
	"errors"
	"net/url"
)

type Challenge struct {
	Id                 int64           `gorm:"column:Id;primaryKey" json:"Id"`
	CreatorId          int64           `gorm:"column:CreatorId" json:"CreatorId"`
	Description        string          `gorm:"column:Description" json:"Description"`
	Name               string          `gorm:"column:Name" json:"Name"`
	Status             ChallengeStatus `gorm:"column:Status" json:"Status"`
	Type               ChallengeType   `gorm:"column:Type" json:"Type"`
	Latitude           float64         `gorm:"column:Latitude" json:"Latitude"`
	Longitude          float64         `gorm:"column:Longitude" json:"Longitude"`
	ExperiencePoints   int             `gorm:"column:ExperiencePoints" json:"ExperiencePoints"`
	KeyPointId         *int64          `gorm:"column:KeyPointId" json:"KeyPointId,omitempty"`
	Image              *url.URL        `gorm:"-" json:"Image,omitempty"`
	LatitudeImage      *float64        `gorm:"column:LatitudeImage" json:"LatitudeImage,omitempty"`
	LongitudeImage     *float64        `gorm:"column:LongitudeImage" json:"LongitudeImage,omitempty"`
	Range              float64         `gorm:"column:Range" json:"Range"`
	RequiredAttendance *int            `gorm:"column:RequiredAttendance" json:"RequiredAttendance,omitempty"`
}

func (Challenge) TableName() string {
	return "encounters.Challenges"
}

func (c *Challenge) Validate() error {
	if c.Description == "" {
		return errors.New("invalid Description")
	}
	if c.Name == "" {
		return errors.New("invalid Name")
	}
	if c.Latitude > 90 || c.Latitude < -90 {
		return errors.New("invalid Latitude")
	}
	if c.Longitude > 180 || c.Longitude < -180 {
		return errors.New("invalid Longitude")
	}
	return nil
}

type ChallengeStatus int

const (
	Draft ChallengeStatus = iota
	Active
	Archived
)

type ChallengeType int

const (
	Social ChallengeType = iota
	Location
	Misc
)
